### ISSUE TITLE
[Feature]: New and Improved Macro Editor

### ISSUE THREAD

**Issue by Tdue21** [RPTools/maptool#4236]
### Describe the Problem

In many ways the current macro editor is fine as it is, and has received multiple updates over time. 
However, if one is working with many macros at a time, it becomes cumbersome to use. 

It quickly becomes annoying that the macro code is on one tab, macro properties on a second, and any kind of tooltip on a third.
Additionally, it is not easy to organise several macro windows in any meaningful way, as they do not figure on the (in my case) Windows task bar. 
They do receive focus along with the main windows, but even with two monitors, they quickly become difficult to work with. 
Additionally, they are on-top of the main windows, so if you need to check something on the map, or library, or whatever, the macro editor is in the way. 

### The Solution you'd like

I would like to propose an entirely new approach to the macro editor dialog. 

First of all, I would like the three tabs to be joined into one, much like a typical IDE, a possible wireframe mock illustrates this: 
![image](https://github.com/RPTools/maptool/assets/2953375/d13c835d-0926-4d47-be39-78b578f793b5)

This macro editor window has 4 main areas: 
* A toolbar at the top. This will have the typical controls and commands, like `Run`, `Save`, maybe `Find` and `Replace`. 
* Some kind of PropertyGrid on the left (or right, I am not picky) for the macro properties and tooltip. 
* A status bar at the bottom with status messages, like save state, maybe tooltip for toolbar commands etc. It is optional.
* A multi-tab editor, with one tab per macro. 

In addition the window should have a separate button on the taskbar, instead of being on-top of the main window. 

By turning the current macro window into a multi-tab IDE like interface, it would be much easier to navigate multiple macros at the same time, and navigate the main window as well. 

### Alternatives that you've considered.

Only real alternatives are external editors, and those present their own set of challenges, like testing.

### Additional Context

My Java knowledge is several decades out of date, but if I was pointed in the right direction, I might be able to design the UI at least. 
If a knowledgeable java developer would take of the code, I would be more than happy to assist with obnoxious questions, testing etc.

**Comment by Tdue21** [comment]
Possibly related issues: 
* #3051
* #3734
* #4200

**Comment by Azhrei** [comment]
As an alternative, would a dockable panel work for your purposes?  Basically the same UI as your mockup, but implemented as a panel?  It could be dragged to another monitor, docked with other panels, hidden, and so on.

It would encapsulate all open macro editors unless the user specifically split one macro tab into its own panel (but some way to move other tabs to that same panel would be needed?).

If that’s acceptable, then the layout becomes an HTML5 template making use of the recent CSS variables that represent themeable components, and the layout can be developed and debugged in a web browser. (And tested there, if the MT scripting kernel were mocked up.)

**Comment by Tdue21** [comment]
That could work perhaps. I'd still like tab 2 & 3 to become part of tab 1, like e.g. a property grid. 
I am sure that there are other nice ways of doing it, but Property grid is the one I can think of :D

**Comment by Tdue21** [comment]
Oh, I think I understand your comment about layout. Are you saying that the layout of the macro editor could be controlled by a html5 template? And thus it is would be fairly easy to customize it?

**Comment by Azhrei** [comment]
Yes, all fields on one “screen”, with tabs for each open macro being edited.

I’m not sure how customizable it would be. On the one hand, it could be a template kept externally like the themes are, but on the other hand it has to interface with the Java code (again, kind of like how theme additions need updates to the Java code). It’s really up to whoever does the implementation.

**Comment by Tdue21** [comment]
Oh yes, the customization part would be implementation dependent, I understand that. It was more of a clarification question.

**Comment by StefanoMarina** [comment]
> Possibly related issues:
> 
>     * [JavaScript, CSS, and HTML syntax highlighting / macro editor modes #3051](https://github.com/RPTools/maptool/issues/3051)
> 
>     * [[Feature]: Macro editor as a dedicated editor #3734](https://github.com/RPTools/maptool/issues/3734)
> 
>     * [[Feature]: Update macro editor title when the macro label is changed #4200](https://github.com/RPTools/maptool/issues/4200)

Yes, what @Tdue21  is requesting is basically what I'm requesting  in request #3734, I'd say this request has better details than mine so we can close mine or merge it into this one.  Fully agree on the proposal, maybe I'd left the buttons on the bottom side to keep consistency.
